kraken-build/src/kraken/std/cargo/manifest.py

```diff
@@ -79,6 +79,15 @@ def read(cls, project_dir: Path) -> CargoMetadata:
             "--manifest-path",
             str(project_dir / "Cargo.toml"),
         ]
+        if locked is None:
+            for parent in (project_dir / "Cargo.toml").absolute().parents:
```

Comment:
I wonder if we should stop at some kind of boundary (max depth, not go beyond the home directory)? Having a rogue `Cargo.toml` lying around could mess this up. (This has happened before; someone had a rogue `pyproject.toml` lying in the home directory that altered the behavior of Poetry (iirc) and it took a while to debug).

Reply:
Yeah, I wanted something like that too, but couldn't really think of any stop heuristics that don't bring their own problems. Stopping at the home directory gets trick cross-platform. We _can_ do it, but would need either a library or heuristics to determine what `$HOME` is. Stopping when we no longer see a `Cargo.toml` is tempting, but would break for workspaces where the members are two levels deep in the hierarchy rather than one without a `Cargo.toml` in between (which is possible). A heuristic like "stop if you've seen two directories without `Cargo.toml`s after seeing at least one `Cargo.tom`, but that gets a) pretty arbitrary, and b) incurs the risk of picking up strays again.

One thing worth highlighting here is that we're looking for `Cargo.lock`, _not_ `Cargo.toml`. And my suspicion is that stray `Cargo.lock`s are much less common than stray `Cargo.toml`s.

Reply:
Yep, I buy those arguments!

One last question: The join with `Cargo.toml` is only to make sure we also check `project_dir`, yes? In that case, I think this is less distracting:

```suggestion
            for parent in (project_dir, *project_dir.parents):
```

Reply:
Yep, that's why, although I think we still need `.absolute()`, because otherwise `.parents` would be empty when `project_dir` is `.` for example. At least my understand of `parents` is that it operates on `dir.split("/")` rather than resolving `..` at each step (but I could be wrong about that).

Also, wouldn't `()` need to become `[]`?

Reply:
Agreed, we should keep it. I accidentally dropped it from the example!

Reply:
Is
```rust
for parent in [project_dir, *project_dir.absolute().parents]:
```
clearer than what's there now though?

Reply:
I think the intent is clearer, yes. Could move the `.absolute()` to a separate line to make it more readable as well

Reply:
```suggestion
            project_dir = project_dir.absolute()
            for parent in [project_dir, *project_dir.parents]:
```